usecases/Concrete/knowledgeGraph/requirements.txt

```diff
@@ -8,5 +8,6 @@ SPARQLWrapper==1.8.5
 requests==2.22.0
```

Comment:
This file is also outdated, the conda environment at the top level is the one and only to be used.

Reply:
We can ignore these commits then, all the libraries I need are already part of the environment file.

Reply:
I've also deleted the validation.py from Concrete

Reply:
For the tests, is there a graph that I can use to test my code? The usecases/MinimumWorkingExample/emodul/processed_data folder is empty but I would need an EM_Graph.ttl file or so to test the validation. I can also just create my own otherwise from the data.

Reply:
Could you talk to @PoNeYvIf  on that? For the tests, it would also be possible (and maybe even better) to create your own very simple test cases (it does not have to be related to a very specific ontology, but rather a very general one such as foaf or prov). In particular make sure that there are errors existing such that those can be returned and processed. We could also have a short phone call on that.

Reply:
@joergfunger Okay, I've added a simple graph and some shapes to test it against. I've added the files to usecases/MinimumWorkingExample/emodul/validation_test/ but I'm not sure if that is the right directory for the test data files. I can move them if another location would be better. I've also added the file tests/validation/test_validation.py which tests the shapes against the graph and checks that some shapes fail and others don't.

Reply:
For the validation, this would work. The only thing remaining before we can merge is the adaption of the doit script in usecases/minimumWorkingExample (thus not dodo.py in Concrete) and the removal of the changes related to the requirements in Concrete.

Reply:
I've removed the changes in requirements, im not sure how the adoption for the doit script should look like, should i talk to @PoNeYvIf or should we schedule a meeting maybe?

Reply:
You can have a look [here](https://github.com/BAMresearch/LebeDigital/blob/main/usecases/MinimumWorkingExample/dodo.py). The final task loops over all the processed data files and generates corresponding ttl files. We would then do another validation task that loops over all the ttl files and validates them. If there are further questions, let's have a short phone call.

Reply:
The only thing that is currently not completly correct in this current dodo file (@PoNeYvIf) is the dependency on the metadatafiles. They are not existing when building the workflow graph and does might lead to problems for the first execution. Thus, there are two options, one is a decorator after (I think @eriktamsen) wanted to change that, and the second option is making the ttl file dependent on the original raw data file that is always present (thus not running into the problem that they are not existing at the very first run).

Reply:
As Jörg said, those two options exist. I successfully used the decorator option in a different branch: https://github.com/BAMresearch/LebeDigital/blob/44-getting-data-from-kg-for-calibration/usecases/MinimumWorkingExample/dodo.py
Note, I also added there the option of a "cheap" workflow as the default, which only processes a single file (we might change it in the future once we can asseble the whole KG), mostly because the new calibration step is too expensive to run every time for every file. If you have questions, let me know.

Reply:
@joergfunger Sorry for the late reply. Okay, so I would loop over all ttl files and check the validation for each, what is the output then? Should I abort the script if one validation fails? Or should I output the results in a text file?

Reply:
The would be that we iteratively generate a knowledgegraph, i.e. we have one large triple store with all the existing triples. Now we would add (in our case this is the loop over all individual ttl files) the data for a new test. @PoNeYvIf is currently trying to do this. Before we actually add each individual file to the triple store, we would check the correctness of the data set, i.e. using the shacl validation. That said, we should have one loop over all raw data files. In this loop, there have to be the substeps 
1. create meta_data.yaml (extract the meta data from the raw data)
2. create knowledge graph of metadata.yaml using the ontology
3. shacl-validate the (not uploaded) knowledge graph for the single experiment
4. if successful, add to global knowledge graph (in reality this would be the triple store, right now it is just our own joint ttl file)
So the subsequent addition to the knowledge graph is only done if it the check is sucessful, otherwise exit the procudure with an error message and a file the logs the shacl error graph.

One thing that I'm not sure how to do that in shacl is the question of checking for instances that are refered to in one graph but defined in another. (e.g. you have a machine whose description is already added to the global triple store, now you would like to add a compression tests which refers to this machine by an id, but we somehow want to test that this id - i.e. an instance of testing machine - is actually existing in the global KG). Is that possible with shacl?

Reply:
So it would make sense to make the task fail if the validation does not succeed so that the next step (which would be the adding to the global KG) is not executed. 

As for the information stored in the global KG, we could read graphs from remote triple stores and join them with local graphs, this would mean some overhead though with copying the global KG every time we want to add to it. If we always check for the same properties with shacl we could write a SPARQL query that gathers the relevant data from the global KG based on the data in the local KG. Maybe @firmao can help?

Reply:
E.g. if we know that we test for machine ids via shacl, we could get the machine IDs from the global KG that are referred to in the local KG. This would of course mean that every time we add a rule we have to think about the SPARQL query and the information that could be contained globally and not locally, i don't know if this is feasible.

Reply:
Yes, let's try to create the queries, let me know when it's okay for you.

Reply:
Not sure if I would always download the complete global graph. It would rather be necessary if particular instances that are referred to in the local graph (e.g. the testing machine that has already be created, or the mix_design we are referring to when adding a new Youngs modulus test) is existing. So for the specific tests, we would first generate the mix design, and upload it to the global KG. Then we would create a Youngs modulus test that references an instance of the mix design (without creating the same instance again). By uploading that then to the KG, these two data sets (mix design and Youngs modulus test) are automatically connected. And before uploading the Youngs modulus KG, we would have to check, if this instance of the mix design is already existing in the global KG (and potentially further properties apart from the id, but that is not necessary right now).

Reply:
As for the sparql query, we somehow now in the local KG generation what classes we expect. So we the sparql query should return the ids of all instances of a given class that should be quite general and would not require rewriting the sparql query each time we have a new test.

Reply:
Maybe we should have a talk together with @firmao to discuss the specifics?